rustls-libssl/src/sign.rs

```diff
@@ -0,0 +1,211 @@
+use rustls::client::ResolvesClientCert;
+use rustls::pki_types::CertificateDer;
+use rustls::sign;
+use rustls::{SignatureAlgorithm, SignatureScheme};
+
+use crate::error;
+use crate::evp_pkey::{
+    rsa_pkcs1_sha256, rsa_pkcs1_sha384, rsa_pkcs1_sha512, rsa_pss_sha256, rsa_pss_sha384,
+    rsa_pss_sha512, EvpPkey, EvpScheme,
+};
+use crate::x509::OwnedX509Stack;
+
+/// This matches up to the implied state machine in `SSL_CTX_use_certificate_chain_file`
+/// and `SSL_CTX_use_PrivateKey_file`, and matching man pages.
+#[derive(Clone, Default)]
+pub struct CertifiedKeySet {
+    /// Last `SSL_CTX_use_certificate_chain_file` result, pending a matching
+    /// `SSL_CTX_use_PrivateKey_file`.
+    pending_cert_chain: Option<Vec<CertificateDer<'static>>>,
+
+    /// Last `SSL_CTX_use_certificate` result, prepended to chain during commit.
+    /// May be absent.
+    pending_cert_end: Option<CertificateDer<'static>>,
+
+    /// The key and certificate we're currently using.
+    ///
+    /// TODO: support multiple key types, and demultiplex them by type.
+    current_key: Option<OpenSslCertifiedKey>,
+}
+
+impl CertifiedKeySet {
+    pub fn stage_certificate_chain(&mut self, chain: Vec<CertificateDer<'static>>) {
+        self.pending_cert_chain = Some(chain);
+    }
+
+    pub fn stage_certificate_end(&mut self, end: CertificateDer<'static>) {
+        self.pending_cert_end = Some(end);
+    }
+
+    pub fn commit_private_key(&mut self, key: EvpPkey) -> Result<(), error::Error> {
+        let chain = match (self.pending_cert_end.take(), self.pending_cert_chain.take()) {
+            (Some(end), Some(mut chain)) => {
+                chain.insert(0, end);
+                chain
+            }
+            (None, Some(chain)) => chain,
+            (Some(end), None) => vec![end],
+            (None, None) => {
+                return Err(error::Error::bad_data("no certificate found for key"));
+            }
+        };
+
+        self.current_key = Some(OpenSslCertifiedKey::new(chain, key));
+        Ok(())
+    }
+
+    pub fn resolver(&self) -> Option<Arc<dyn ResolvesClientCert>> {
+        self.current_key.as_ref().map(|ck| ck.resolver())
+    }
+
+    /// For `SSL_get_certificate`
+    pub fn borrow_current_cert(&self) -> *mut X509 {
+        self.current_key
+            .as_ref()
+            .map(|ck| ck.borrow_cert())
+            .unwrap_or(ptr::null_mut())
+    }
+
+    /// For `SSL_get_privatekey`
+    pub fn borrow_current_key(&self) -> *mut EVP_PKEY {
+        self.current_key
+            .as_ref()
+            .map(|ck| ck.borrow_key())
+            .unwrap_or(ptr::null_mut())
+    }
+}
+
+#[derive(Clone)]
+struct OpenSslCertifiedKey {
+    key: EvpPkey,
+    openssl_chain: OwnedX509Stack,
+    rustls_chain: Vec<CertificateDer<'static>>,
+}
+
+impl OpenSslCertifiedKey {
+    fn new(chain: Vec<CertificateDer<'static>>, key: EvpPkey) -> Self {
+        let openssl_chain = OwnedX509Stack::from_rustls(&chain);
```

Comment:
nit: Feels tidier to inline this into `Self` initialization since the binding isn't used elsewhere and `rustls_chain` already requires an explicit argument.